README
DESCRIPTION
    LZF is an extremely fast (not that much slower than a pure memcpy)
    compression algorithm. It is ideal for applications where you want to
    save *some* space but not at the cost of speed. It is ideal for
    repetitive data as well. The module is self-contained and very small (no
    large library to be pulled in).

    I do not know for certain wether any patents in any countries apply
    to this algorithm, but at the moment it is believed that it is free
    from any patents. More importantly, it is also free to use in every
    software package (see LICENSE).

    See the lzf.h file for details on how the functions in this
    mini-library are to be used.

    NOTE: This package contains a very bare-bones command-line utility
    which is neither optimized for speed nor for compression. This library
    is really intented to be used inside larger programs.

AUTHOR
    This library was written by Marc Lehmann <pcg@goof.com> (See also
    http://liblzf.plan9.de/).




Changes

TODO: try unaligned copy again in decompressor
TODO: allow size-optimised binaries by avoiding unrolling
TODO: implement lzf_c_best in lzf.
TODO: fix lzf_c_best.

3.7 (unreleased)
	- add lzf_c_best.c, a slower but better compressor.
	- switch to a multiplicative hash (developed with Steinar Gunderson),
          which is faster on modern cpus and compresses a bit better. The old
          hash function which uses only shifts is still available.
        - allow user configurable hash table slots, which makes it possible
          to use e.g. 16 bit offsets for a smaller hashtable (if your data is
          always < 64kb).
	- use _WIN32, not WIN32, when testing for windows (fails with bcc),
          patch by Tamas Tevesz.
        - try to port to win64+gcc, which is non-posix but doesn't have _int64
          (reported by Maciej Adamczyk).

3.6  Mon Feb  7 17:37:31 CET 2011
	- fixed hash calculation in C♯ version (Tiago Freitas Leal).
        - unroll copy for small sizes, use memcpy for larger sizes,
          greatly speeding up decompression in most cases.
        - finally disable rep movsb - it's a big loss on modern intel cpus,
          and only a small win on amd cpus.
        - improve C++ compatibility of the code.
        - slightly improve compressor speed.
        - halved memory requirements for compressor on 64 bit architectures,
          which can improve the speed quite a bit on older cpus.

3.5  Fri May  1 02:28:42 CEST 2009
	- lzf_compress did sometimes write one octet past the given output
          buffer (analyzed and nice testcase by Salvatore Sanfilippo).

3.4  Tue Sep  2 06:45:00 CEST 2008
	- the fix from 3.3 introduced a compression bug, which is fixed in
          this release (which explains the mysterious prerelease...). Thanks
          once more to Clément Calmels.

3.3  Mon Aug 25 03:17:42 CEST 2008
	- lzf_compress could access memory after the given input buffer
          when outputting back references. reported with nice testcase
          by Clément Calmels.

3.2  Fri May  9 18:52:23 CEST 2008
	- include a workaround for failing POSIX and real-world compliance
          on 64 bit windows (microsoft claims to support POSIX, but is far
          from it). (bug found and analysed nicely by John Lilley).

3.1  Fri Nov 30 11:33:04 CET 2007
	- IMPORTANT BUGFIX: a too long final literal run would corrupt data
          in the encoder (this was introduced in 3.0 only, earlier versions
          are safe).

3.0  Tue Nov 13 22:13:09 CET 2007
	- switched to 2-clause bsd with "GPL v2 or any later version" option.
        - speed up compression by ~10-15% in common cases
          by some manual unrolling.
        - import some compiler tricks from JSON::XS, for further speed-ups.
        - tune hash functions depending on ULTRA_FAST or VERY_FAST settings.
        - for typical binary data (e.g. /bin/bash, memory dumps,
          canterbury corpus etc.), speed is now comparable to fastlz, but
          with better compression ratio. with ULTRA_FAST, it's typically
          3-15% faster than fastlz while still maintaining a similar ratio.
          (amd64 and core 2 duo, ymmv). thanks a lot for the competition :)
        - undo inline assembly in compressor, it is no longer helpful.
        - no changes to the decompressor.
        - use a HLOG of 16 by default now (formerly 15).

2.1  Fri Nov  2 13:34:42 CET 2007
	- switched to a 2-clause bsd license with GPL exception.
	- get rid of memcpy.
        - tentatively use rep movsb on x86 and x86_64 (gcc only) for a
          moderate speed improvement.
        - applied patch by Kein-Hong Man to maske lzf.c compile under
          the crippled mingw32 environment.

2.0  Fri Feb 16 23:11:18 CET 2007
	- replaced lzf demo by industrial-strength lzf utility with behaviour
          similar other compression utilities. Thanks for Stefan Traby for
          rewriting it!
        - fix state arg prototype.

1.7  Wed Sep 27 17:29:15 CEST 2006
	- remove bogus "unlzf" patch.
          note to self: never accept well-meant patches.
	- make lzf more robust in presence of padding bytes or sudden eof.

1.6  Fri Jul  7 17:31:26 CEST 2006
	- the lzf example utility will now uncompress if invoked
          as "unlzf" (patch by Scott Feeney).
        - add CHECK_INPUT option that adds more checks for input
          data validity.
        - help applications that do not pass in the correct length
          (such as php) by returning either EINVAL or E2BIG.
        - default HLOG size is now 15 (cpu caches have increased).
        - documentation fixes.

1.51 Thu Apr 14 22:15:46 CEST 2005
	- incorporated C♯ implementation of both the en- and decoder,
          written by "Oren J. Maurice".
          You can find it in the cs/ subdirectory.
        - make FRST, NEXT IDX overridable if lzf_c.c is directly included
          in the code.

1.5  Tue Mar  8 20:23:23 CET 2005
	- incorporated improvements by Adam D. Moss,
          which includes a new VERY_FAST mode which is
          a bit slower than ULTRA_FAST but much better,
          and enabled it as default.

1.401 Thu Mar  3 18:00:52 CET 2005
	- use cstring in c++, not string.h.
        - change of contact address.

1.4  Wed Dec 15 08:08:49 CET 2004
	- very very slight tuning of the hashing function.

1.3  Thu Mar 25 15:41:17 CET 2004
	- changed license of lzf core code to explicitly allow
          relicensing under the GPLv2.
	- added VPATH support as suggested by Björn Eriksson.

1.2  Mon Dec 29 13:47:28 CET 2003
	- avoid spurious memory accesses after the to-be-compressed
          memory region. originally reported by Michal Zalewski.
        - flip LZF_STACK_ARG meaning (to be correct).

1.1  Tue Dec 23 05:48:32 CET 2003
	- removed #warn directive, it's not worth the hassle.
        - add LZF_STACK_ARG and AVOID_ERRNO configurations
          for embedded systems.
        - make it compile cleanly as c++.
        - some small documentation and code fixes.


lzf.c
/*
 * Copyright (c) 2006      Stefan Traby <stefan@hello-penguin.com>
 * Copyright (c) 2012      Marc Lehmann <schmorp@schmorp.de>
 * 
 * Redistribution and use in source and binary forms, with or without modifica-
 * tion, are permitted provided that the following conditions are met:
 * 
 *   1.  Redistributions of source code must retain the above copyright notice,
 *       this list of conditions and the following disclaimer.
 * 
 *   2.  Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MER-
 * CHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
 * EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPE-
 * CIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTH-
 * ERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * Alternatively, the contents of this file may be used under the terms of
 * the GNU General Public License ("GPL") version 2 or any later version,
 * in which case the provisions of the GPL are applicable instead of
 * the above. If you wish to allow the use of your version of this file
 * only under the terms of the GPL and not to allow others to use your
 * version of this file under the BSD license, indicate your decision
 * by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL. If you do not delete the
 * provisions above, a recipient may use your version of this file under
 * either the BSD or the GPL.
 */

#include "config.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <limits.h>
#include "lzf.h"
#include "lzf_c_best.c"

#ifdef HAVE_GETOPT_H
# include <getopt.h>
#endif

#define BLOCKSIZE (1024 * 64 - 1)
#define MAX_BLOCKSIZE BLOCKSIZE

typedef unsigned char u8;

static off_t nr_read, nr_written;

static const char *imagename;
static enum { compress, uncompress, lzcat } mode = compress;
static int verbose = 0;
static int best = 0;
static int force = 0;
static long blocksize = BLOCKSIZE;

#ifdef HAVE_GETOPT_LONG

  struct option longopts[] = {
    {"compress"  , 0, 0, 'c'},
    {"decompress", 0, 0, 'd'},
    {"uncompress", 0, 0, 'd'},
    {"best"      , 0, 0, '9'},
    {"force"     , 0, 0, 'f'},
    {"help"      , 0, 0, 'h'},
    {"verbose"   , 0, 0, 'v'},
    {"blocksize" , 1, 0, 'b'},
    {0           , 0, 0,   0}
  };

  static const char *opt =
    "-c --compress    compress\n"
    "-d --decompress  decompress\n"
    "-9 --best        best compression\n"
    "-f --force       force overwrite of output file\n"
    "-h --help        give this help\n"
    "-v --verbose     verbose mode\n"
    "-b # --blocksize # set blocksize\n"
    "\n";

#else

  static const char *opt =
    "-c   compress\n"
    "-d   decompress\n"
    "-9   best compression\n"
    "-f   force overwrite of output file\n"
    "-h   give this help\n"
    "-v   verbose mode\n"
    "-b # set blocksize\n"
    "\n";

#endif

static void
usage (int rc)
{
  fprintf (stderr, "\n"
           "lzf, a very lightweight compression/decompression utility written by Stefan Traby.\n"
           "uses liblzf written by Marc Lehmann <schmorp@schmorp.de> You can find more info at\n"
           "http://liblzf.plan9.de/\n"
           "\n"
           "usage: lzf [-dufhvb9] [file ...]\n"
           "       unlzf [file ...]\n"
           "       lzcat [file ...]\n"
           "\n%s",
           opt);

  exit (rc);
}

static inline ssize_t
rread (int fd, void *buf, size_t len)
{
  ssize_t rc = 0, offset = 0;
  char *p = buf;

  while (len && (rc = read (fd, &p[offset], len)) > 0)
    {
      offset += rc;
      len -= rc;
    }

  nr_read += offset;

  if (rc < 0)
    return rc;

  return offset;
}

/* returns 0 if all written else -1 */
static inline ssize_t
wwrite (int fd, void *buf, size_t len)
{
  ssize_t rc;
  char *b = buf;
  size_t l = len;

  while (l)
    {
      rc = write (fd, b, l);
      if (rc < 0)
        {
          fprintf (stderr, "%s: write error: ", imagename);
          perror ("");
          return -1;
        }

      l -= rc;
      b += rc;
    }

  nr_written += len;
  return 0;
}

/*
 * Anatomy: an lzf file consists of any number of blocks in the following format:
 *
 * \x00   EOF (optional)
 * "ZV\0" 2-byte-usize <uncompressed data>
 * "ZV\1" 2-byte-csize 2-byte-usize <compressed data>
 * "ZV\2" 4-byte-crc32-0xdebb20e3 (NYI)
 */


#define TYPE0_HDR_SIZE 5
#define TYPE1_HDR_SIZE 7
#define MAX_HDR_SIZE 7
#define MIN_HDR_SIZE 5

static int
compress_fd (int from, int to)
{
  ssize_t us, cs, len;
  u8 buf1[MAX_BLOCKSIZE + MAX_HDR_SIZE + 16];
  u8 buf2[MAX_BLOCKSIZE + MAX_HDR_SIZE + 16];
  u8 *header;

  nr_read = nr_written = 0;
  while ((us = rread (from, &buf1[MAX_HDR_SIZE], blocksize)) > 0)
    {
      cs = (best ? lzf_compress_best : lzf_compress) (&buf1[MAX_HDR_SIZE], us, &buf2[MAX_HDR_SIZE], us > 4 ? us - 4 : us);
      if (cs)
        {
          header = &buf2[MAX_HDR_SIZE - TYPE1_HDR_SIZE];
          header[0] = 'Z';
          header[1] = 'V';
          header[2] = 1;
          header[3] = cs >> 8;
          header[4] = cs & 0xff;
          header[5] = us >> 8;
          header[6] = us & 0xff;
          len = cs + TYPE1_HDR_SIZE;
        }
      else
        {                       // write uncompressed
          header = &buf1[MAX_HDR_SIZE - TYPE0_HDR_SIZE];
          header[0] = 'Z';
          header[1] = 'V';
          header[2] = 0;
          header[3] = us >> 8;
          header[4] = us & 0xff;
          len = us + TYPE0_HDR_SIZE;
        }

      if (wwrite (to, header, len) == -1)
        return -1;
    }

  return 0;
}

static int
uncompress_fd (int from, int to)
{
  u8 header[MAX_HDR_SIZE];
  u8 buf1[MAX_BLOCKSIZE + MAX_HDR_SIZE + 16];
  u8 buf2[MAX_BLOCKSIZE + MAX_HDR_SIZE + 16];
  u8 *p;
  int l, rd;
  ssize_t rc, cs, us, bytes, over = 0;

  nr_read = nr_written = 0;
  while (1)
    {
      rc = rread (from, header + over, MAX_HDR_SIZE - over);
      if (rc < 0)
        {
          fprintf (stderr, "%s: read error: ", imagename);
          perror ("");
          return -1;
        }

      rc += over;
      over = 0;
      if (!rc || header[0] == 0)
        return 0;

      if (rc < MIN_HDR_SIZE || header[0] != 'Z' || header[1] != 'V')
        {
          fprintf (stderr, "%s: invalid data stream - magic not found or short header\n", imagename);
          return -1;
        }

      switch (header[2])
        {
          case 0:
            cs = -1;
            us = (header[3] << 8) | header[4];
            p = &header[TYPE0_HDR_SIZE];
            break;
          case 1:
            if (rc < TYPE1_HDR_SIZE)
              {
                goto short_read;
              }
            cs = (header[3] << 8) | header[4];
            us = (header[5] << 8) | header[6];
            p = &header[TYPE1_HDR_SIZE];
            break;
          default:
            fprintf (stderr, "%s: unknown blocktype\n", imagename);
            return -1;
        }

      bytes = cs == -1 ? us : cs;
      l = &header[rc] - p;

      if (l > 0)
        memcpy (buf1, p, l);

      if (l > bytes)
        {
          over = l - bytes;
          memmove (header, &p[bytes], over);
        }

      p = &buf1[l];
      rd = bytes - l;
      if (rd > 0)
        if ((rc = rread (from, p, rd)) != rd)
          goto short_read;

      if (cs == -1)
        {
          if (wwrite (to, buf1, us))
            return -1;
        }
      else
        {
          if (lzf_decompress (buf1, cs, buf2, us) != us)
            {
              fprintf (stderr, "%s: decompress: invalid stream - data corrupted\n", imagename);
              return -1;
            }

          if (wwrite (to, buf2, us))
            return -1;
        }
    }

  return 0;

short_read:
  fprintf (stderr, "%s: short data\n", imagename);
  return -1;
}

static int
open_out (const char *name)
{
  int fd;
  int m = O_EXCL;

  if (force)
    m = 0;

  fd = open (name, O_CREAT | O_WRONLY | O_TRUNC | m, 600);
#if defined(__MINGW32__)
  _setmode(fd, _O_BINARY);
#endif
  return fd;
}

static int
compose_name (const char *fname, char *oname)
{
  char *p;

  if (mode == compress)
    {
      if (strlen (fname) > PATH_MAX - 4)
        {
          fprintf (stderr, "%s: %s.lzf: name too long", imagename, fname);
          return -1;
        }

      strcpy (oname, fname);
      strcat (oname, ".lzf");
    }
  else
    {
      if (strlen (fname) > PATH_MAX)
        {
          fprintf (stderr, "%s: %s: name too long\n", imagename, fname);
          return -1;
        }

      strcpy (oname, fname);
      p = &oname[strlen (oname)] - 4;
      if (p < oname || strcmp (p, ".lzf"))
        {
          fprintf (stderr, "%s: %s: unknown suffix\n", imagename, fname);
          return -1;
        }

      *p = 0;
    }

  return 0;
}

static int
run_file (const char *fname)
{
  int fd, fd2;
  int rc;
  struct stat mystat;
  char oname[PATH_MAX + 1];

  if (mode != lzcat)
    if (compose_name (fname, oname))
      return -1;

#if !defined(__MINGW32__)
  rc = lstat (fname, &mystat);
#else
  rc = stat (fname, &mystat);
#endif
  fd = open (fname, O_RDONLY);
#if defined(__MINGW32__)
  _setmode(fd, _O_BINARY);
#endif
  if (rc || fd == -1)
    {
      fprintf (stderr, "%s: %s: ", imagename, fname);
      perror ("");
      return -1;
    }

  if (!S_ISREG (mystat.st_mode))
    {
      fprintf (stderr, "%s: %s: not a regular file.\n", imagename, fname);
      close (fd);
      return -1;
    }

  if (mode == lzcat)
    {
      rc = uncompress_fd (fd, 1);
      close (fd);
      return rc;
    }

  fd2 = open_out (oname);
  if (fd2 == -1)
    {
      fprintf (stderr, "%s: %s: ", imagename, oname);
      perror ("");
      close (fd);
      return -1;
    }

  if (mode == compress)
    {
      rc = compress_fd (fd, fd2);
      if (!rc && verbose)
        fprintf (stderr, "%s:  %5.1f%% -- replaced with %s\n",
                 fname, nr_read == 0 ? 0 : 100.0 - nr_written / ((double) nr_read / 100.0), oname);
    }
  else
    {
      rc = uncompress_fd (fd, fd2);
      if (!rc && verbose)
        fprintf (stderr, "%s:  %5.1f%% -- replaced with %s\n",
                 fname, nr_written == 0 ? 0 : 100.0 - nr_read / ((double) nr_written / 100.0), oname);
    }

#if !defined(__MINGW32__)
  fchmod (fd2, mystat.st_mode);
#else
  chmod (oname, mystat.st_mode);
#endif
  close (fd);
  close (fd2);

  if (!rc)
    unlink (fname);

  return rc;
}

int
main (int argc, char *argv[])
{
  char *p = argv[0];
  int optc;
  int rc = 0;

  errno = 0;
  p = getenv ("LZF_BLOCKSIZE");
  if (p)
    {
      blocksize = strtoul (p, 0, 0);
      if (errno || !blocksize || blocksize > MAX_BLOCKSIZE)
        blocksize = BLOCKSIZE;
    }

  p = strrchr (argv[0], '/');
  imagename = p ? ++p : argv[0];

  if (!strncmp (imagename, "un", 2) || !strncmp (imagename, "de", 2))
    mode = uncompress;

  if (strstr (imagename, "cat"))
    mode = lzcat;

#ifdef HAVE_GETOPT_LONG
  while ((optc = getopt_long (argc, argv, "cd9fhvb:", longopts, 0)) != -1)
#else
  while ((optc = getopt (argc, argv, "cd9fhvb:")) != -1)
#endif
    {
      switch (optc)
        {
          case 'c':
            mode = compress;
            break;
          case 'd':
            mode = uncompress;
            break;
          case '9':
            best = 1;
            break;
          case 'f':
            force = 1;
            break;
          case 'h':
            usage (0);
            break;
          case 'v':
            verbose = 1;
            break;
          case 'b':
            errno = 0;
            blocksize = strtoul (optarg, 0, 0);
            if (errno || !blocksize || blocksize > MAX_BLOCKSIZE)
              blocksize = BLOCKSIZE;
            break;
          default:
            usage (1);
            break;
        }
    }

  if (optind == argc)
    {                           // stdin stdout
      if (!force)
        {
          if ((mode == uncompress || mode == lzcat) && isatty (0))
            {
              fprintf (stderr, "%s: compressed data not read from a terminal. Use -f to force decompression.\n", imagename);
              exit (1);
            }
          if (mode == compress && isatty (1))
            {
              fprintf (stderr, "%s: compressed data not written to a terminal. Use -f to force compression.\n", imagename);
              exit (1);
            }
        }

      if (mode == compress)
        rc = compress_fd (0, 1);
      else
        rc = uncompress_fd (0, 1);

      exit (rc ? 1 : 0);
    }

  while (optind < argc)
    rc |= run_file (argv[optind++]);

  exit (rc ? 1 : 0);
}



lzf.h
/*
 * Copyright (c) 2000-2008 Marc Alexander Lehmann <schmorp@schmorp.de>
 *
 * Redistribution and use in source and binary forms, with or without modifica-
 * tion, are permitted provided that the following conditions are met:
 *
 *   1.  Redistributions of source code must retain the above copyright notice,
 *       this list of conditions and the following disclaimer.
 *
 *   2.  Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MER-
 * CHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
 * EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPE-
 * CIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTH-
 * ERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * Alternatively, the contents of this file may be used under the terms of
 * the GNU General Public License ("GPL") version 2 or any later version,
 * in which case the provisions of the GPL are applicable instead of
 * the above. If you wish to allow the use of your version of this file
 * only under the terms of the GPL and not to allow others to use your
 * version of this file under the BSD license, indicate your decision
 * by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL. If you do not delete the
 * provisions above, a recipient may use your version of this file under
 * either the BSD or the GPL.
 */

#ifndef LZF_H
#define LZF_H

/***********************************************************************
**
**	lzf -- an extremely fast/free compression/decompression-method
**	http://liblzf.plan9.de/
**
**	This algorithm is believed to be patent-free.
**
***********************************************************************/

/* API version (major * 256 + minor)
 * major API version gets bumped on incompatible changes.
 * minor API version gets bumped on compatible changes.
 * 1.5 => 1.6: add LZF_MAX_COMPRESSED_SIZE
 */
#define LZF_VERSION 0x0106

/*
 * Compress in_len bytes stored at the memory block starting at
 * in_data and write the result to out_data, up to a maximum length
 * of out_len bytes.
 *
 * If the output buffer is not large enough or any error occurs return 0,
 * otherwise return the number of bytes used, which might be considerably
 * more than in_len (but less than 1 + 104% of the original size), so it
 * makes sense to always use out_len == in_len - 1), to ensure _some_
 * compression, and store the data uncompressed otherwise (with a flag, of
 * course.
 *
 * lzf_compress might use different algorithms on different systems and
 * even different runs, thus might result in different compressed strings
 * depending on the phase of the moon or similar factors. However, all
 * these strings are architecture-independent and will result in the
 * original data when decompressed using lzf_decompress.
 *
 * The buffers must not be overlapping.
 *
 * If the option LZF_STATE_ARG is enabled, an extra argument must be
 * supplied which is not reflected in this header file. Refer to lzfP.h
 * and lzf_c.c.
 *
 */
unsigned int
lzf_compress (const void *const in_data,  unsigned int in_len,
              void             *out_data, unsigned int out_len);

/*
 * The maximum out_len that needs to be allocated to make sure
 * any input data can be compressed without overflowing the output
 * buffer, i.e. maximum out_len = LZF_MAX_COMPRESSED_SIZE (in_len).
 * This is useful if you don't want to bother with the case of
 * incompressible data and just want to provide a buffer that is
 * guaranteeed to be big enough.
 * This macro can be used at preprocessing time.
 */
#define LZF_MAX_COMPRESSED_SIZE(n) ((n) * 33 / 32 + 1)

/*
 * Decompress data compressed with some version of the lzf_compress
 * function and stored at location in_data and length in_len. The result
 * will be stored at out_data up to a maximum of out_len characters.
 *
 * If the output buffer is not large enough to hold the decompressed
 * data, a 0 is returned and errno is set to E2BIG. Otherwise the number
 * of decompressed bytes (i.e. the original length of the data) is
 * returned.
 *
 * If an error in the compressed data is detected, a zero is returned and
 * errno is set to EINVAL.
 *
 * This function is very fast, about as fast as a copying loop.
 */
unsigned int
lzf_decompress (const void *const in_data,  unsigned int in_len,
                void             *out_data, unsigned int out_len);

#endif



lzf_d.c
/*
 * Copyright (c) 2000-2010 Marc Alexander Lehmann <schmorp@schmorp.de>
 *
 * Redistribution and use in source and binary forms, with or without modifica-
 * tion, are permitted provided that the following conditions are met:
 *
 *   1.  Redistributions of source code must retain the above copyright notice,
 *       this list of conditions and the following disclaimer.
 *
 *   2.  Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MER-
 * CHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
 * EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPE-
 * CIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTH-
 * ERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * Alternatively, the contents of this file may be used under the terms of
 * the GNU General Public License ("GPL") version 2 or any later version,
 * in which case the provisions of the GPL are applicable instead of
 * the above. If you wish to allow the use of your version of this file
 * only under the terms of the GPL and not to allow others to use your
 * version of this file under the BSD license, indicate your decision
 * by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL. If you do not delete the
 * provisions above, a recipient may use your version of this file under
 * either the BSD or the GPL.
 */

#include "lzfP.h"

#if AVOID_ERRNO
# define SET_ERRNO(n)
#else
# include <errno.h>
# define SET_ERRNO(n) errno = (n)
#endif

#if USE_REP_MOVSB /* small win on amd, big loss on intel */
#if (__i386 || __amd64) && __GNUC__ >= 3
# define lzf_movsb(dst, src, len)                \
   asm ("rep movsb"                              \
        : "=D" (dst), "=S" (src), "=c" (len)     \
        :  "0" (dst),  "1" (src),  "2" (len));
#endif
#endif

unsigned int 
lzf_decompress (const void *const in_data,  unsigned int in_len,
                void             *out_data, unsigned int out_len)
{
  u8 const *ip = (const u8 *)in_data;
  u8       *op = (u8 *)out_data;
  u8 const *const in_end  = ip + in_len;
  u8       *const out_end = op + out_len;

  do
    {
      unsigned int ctrl = *ip++;

      if (ctrl < (1 << 5)) /* literal run */
        {
          ctrl++;

          if (op + ctrl > out_end)
            {
              SET_ERRNO (E2BIG);
              return 0;
            }

#if CHECK_INPUT
          if (ip + ctrl > in_end)
            {
              SET_ERRNO (EINVAL);
              return 0;
            }
#endif

#ifdef lzf_movsb
          lzf_movsb (op, ip, ctrl);
#else
          switch (ctrl)
            {
              case 32: *op++ = *ip++; case 31: *op++ = *ip++; case 30: *op++ = *ip++; case 29: *op++ = *ip++;
              case 28: *op++ = *ip++; case 27: *op++ = *ip++; case 26: *op++ = *ip++; case 25: *op++ = *ip++;
              case 24: *op++ = *ip++; case 23: *op++ = *ip++; case 22: *op++ = *ip++; case 21: *op++ = *ip++;
              case 20: *op++ = *ip++; case 19: *op++ = *ip++; case 18: *op++ = *ip++; case 17: *op++ = *ip++;
              case 16: *op++ = *ip++; case 15: *op++ = *ip++; case 14: *op++ = *ip++; case 13: *op++ = *ip++;
              case 12: *op++ = *ip++; case 11: *op++ = *ip++; case 10: *op++ = *ip++; case  9: *op++ = *ip++;
              case  8: *op++ = *ip++; case  7: *op++ = *ip++; case  6: *op++ = *ip++; case  5: *op++ = *ip++;
              case  4: *op++ = *ip++; case  3: *op++ = *ip++; case  2: *op++ = *ip++; case  1: *op++ = *ip++;
            }
#endif
        }
      else /* back reference */
        {
          unsigned int len = ctrl >> 5;

          u8 *ref = op - ((ctrl & 0x1f) << 8) - 1;

#if CHECK_INPUT
          if (ip >= in_end)
            {
              SET_ERRNO (EINVAL);
              return 0;
            }
#endif
          if (len == 7)
            {
              len += *ip++;
#if CHECK_INPUT
              if (ip >= in_end)
                {
                  SET_ERRNO (EINVAL);
                  return 0;
                }
#endif
            }

          ref -= *ip++;

          if (op + len + 2 > out_end)
            {
              SET_ERRNO (E2BIG);
              return 0;
            }

          if (ref < (u8 *)out_data)
            {
              SET_ERRNO (EINVAL);
              return 0;
            }

#ifdef lzf_movsb
          len += 2;
          lzf_movsb (op, ref, len);
#else
          switch (len)
            {
              default:
                len += 2;

                if (op >= ref + len)
                  {
                    /* disjunct areas */
                    memcpy (op, ref, len);
                    op += len;
                  }
                else
                  {
                    /* overlapping, use octte by octte copying */
                    do
                      *op++ = *ref++;
                    while (--len);
                  }

                break;

              case 9: *op++ = *ref++;
              case 8: *op++ = *ref++;
              case 7: *op++ = *ref++;
              case 6: *op++ = *ref++;
              case 5: *op++ = *ref++;
              case 4: *op++ = *ref++;
              case 3: *op++ = *ref++;
              case 2: *op++ = *ref++;
              case 1: *op++ = *ref++;
              case 0: *op++ = *ref++; /* two octets more */
                      *op++ = *ref++;
            }
#endif
        }
    }
  while (ip < in_end);

  return op - (u8 *)out_data;
}

